lib/ex_aws/s3.ex

```diff
@@ -488,22 +489,54 @@ defmodule ExAws.S3 do
 
   Handles initialization, uploading parts concurrently, and multipart upload completion.
 
-  ## Examples
+  ## Uploading a stream
+
+  Streams that emit binaries may be uploaded directly to S3. Each binary will be uploaded
+  as a chunk, so it must be at least 5 megabytes in size. The `S3.Upload.stream_file`
+  helper takes care of reading the file in 5 megabyte chunks.
   ```
   "path/to/big/file"
-  |> S3.Upload.stream_file!
+  |> S3.Upload.stream_file
   |> S3.upload("my-bucket", "path/on/s3")
-  |> ExAws.request! #=> {:ok, :done}
+  |> ExAws.request! #=> :done
   ```
 
-  ## Notes
+  ## Uploading a flow
+
+  GenStage Flows can also be uploaded directly to S3. However, due to the concurrent
+  nature of flows, the producer must emit events in the following format
+  `{binary, index}` to ensure they are uploaded in the correct order. Binaries
+  will be uploaded as a chunk and must be at least 5 megabytes in size.
+  ```
+  enumerable = ["hello world"] |> Enum.with_index(1)
 
-  Source is expected to be an enumerable that emits binaries. Each binary will be
-  uploaded as a chunk, so it must be at least 5 megabytes in size.
+  Flow.new(stages: 4, max_demand: 2)
+  |> Flow.from_enumerable(enumerable)
+  |> S3.upload("my-bucket", "test.txt")
+  |> ExAws.request! #=> :done
+  ```
   """
-  def upload(source, bucket, path, opts) do
+  def upload(source, bucket, path, opts \\ []) do
+    do_upload(source, bucket, path, opts)
+  end
```

Comment:
`do_upload` largely unnecessary. You could just have
```elixir
def upload(source, bucket, path, opts \\ [])
def upload(%Flow{} = flow, bucket, path, opts) do
  # flow stuff
end
def upload(stream, bucket, path, opts) do
  # non flow stuff
end
```

We should also add typespecs. In this case that should be:
```elixir
@spec upload(source :: Enumerable.t | Flow.t, bucket :: binary, path :: binary) :: __MODULE__.Upload.t
@spec upload(source :: Enumerable.t | Flow.t, bucket :: binary, path :: binary, opts :: Keyword.t) :: __MODULE__.Upload.t
```